File: backend/routers/patients.py
"""
routers/patients.py — Patient CRUD endpoints with role-based access control.
"""

from datetime import date
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from database import get_db
from models import Patient, Appointment, User
from schemas import PatientCreate, PatientUpdate, PatientOut
from dependencies import get_current_user, require_role

router = APIRouter(prefix="/patients", tags=["Patients"])


@router.get("", response_model=list[PatientOut])
def get_patients(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get all patients. Accessible by all authenticated roles."""
    return db.query(Patient).order_by(Patient.id).all()


@router.post("", response_model=PatientOut, status_code=status.HTTP_201_CREATED)
def create_patient(
    data: PatientCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_role("admin", "receptionist")),
):
    """
    Register a new patient.
    Accessible by: Admin, Receptionist.
    """
    # Duplicate check: same name + phone
    existing = db.query(Patient).filter(
        Patient.name == data.name.strip(),
        Patient.phone == data.phone.strip(),
    ).first()
    if existing:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="A patient with this name and phone already exists",
        )

    patient = Patient(
        name=data.name.strip(),
        age=data.age,
        gender=data.gender,
        blood=data.blood,
        phone=data.phone.strip(),
        email=(data.email or "").strip(),
        address=(data.address or "").strip(),
        condition=data.condition or "General",
        status=data.status or "Active",
        joined=date.today().isoformat(),
    )
    db.add(patient)
    db.commit()
    db.refresh(patient)
    return patient


@router.put("/{patient_id}", response_model=PatientOut)
def update_patient(
    patient_id: int,
    data: PatientUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_role("admin")),
):
    """
    Update an existing patient record.
    Accessible by: Admin only.
    """
    patient = db.query(Patient).filter(Patient.id == patient_id).first()
    if not patient:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Patient not found")

    update_data = data.model_dump(exclude_unset=True)

    # Strip strings
    for field in ("name", "phone", "email", "address"):
        if field in update_data and update_data[field] is not None:
            update_data[field] = update_data[field].strip()

    # Duplicate check (exclude self)
    if "name" in update_data or "phone" in update_data:
        check_name = update_data.get("name", patient.name)
        check_phone = update_data.get("phone", patient.phone)
        dup = db.query(Patient).filter(
            Patient.id != patient_id,
            Patient.name == check_name,
            Patient.phone == check_phone,
        ).first()
        if dup:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="A patient with this name and phone already exists",
            )

    for key, value in update_data.items():
        setattr(patient, key, value)

    db.commit()
    db.refresh(patient)
    return patient


@router.delete("/{patient_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_patient(
    patient_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_role("admin")),
):
    """
    Delete a patient and cascade-delete their appointments.
    Accessible by: Admin only.
    """
    patient = db.query(Patient).filter(Patient.id == patient_id).first()
    if not patient:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Patient not found")

    # Delete associated appointments first
    db.query(Appointment).filter(Appointment.patient_id == patient_id).delete()
    db.delete(patient)
    db.commit()

File: backend/routes/patients.js
const express = require('express');
const router = express.Router();
const Patient = require('../models/Patient');
const Appointment = require('../models/Appointment');
const { authenticateJWT, requireRole } = require('../middleware/auth');

// Get all patients - All authenticated roles
router.get('/', authenticateJWT, async (req, res) => {
  try {
    const list = await Patient.find({}).sort({ name: 1 });
    res.json(list);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: true, detail: 'Internal server error' });
  }
});

// Create new patient - Admin & Receptionist
router.post('/', authenticateJWT, requireRole('admin', 'receptionist'), async (req, res) => {
  const { name, age, gender, blood, phone, email, address, condition, status } = req.body;

  if (!name || !age || !phone) {
    return res.status(422).json({ error: true, detail: 'Missing required fields: name, age, phone' });
  }

  try {
    // Duplicate check: same name and phone
    const existing = await Patient.findOne({
      name: name.trim(),
      phone: phone.trim(),
    });

    if (existing) {
      return res.status(409).json({
        error: true,
        detail: 'A patient with this name and phone already exists',
      });
    }

    const patient = new Patient({
      name: name.trim(),
      age: parseInt(age, 10),
      gender,
      blood,
      phone: phone.trim(),
      email: (email || '').trim(),
      address: (address || '').trim(),
      condition: condition || 'General',
      status: status || 'Active',
      joined: new Date().toISOString().slice(0, 10),
    });

    await patient.save();
    res.status(201).json(patient);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: true, detail: 'Internal server error' });
  }
});

// Update patient - Admin only
router.put('/:id', authenticateJWT, requireRole('admin'), async (req, res) => {
  const { id } = req.params;
  const updateFields = req.body;

  try {
    const patient = await Patient.findById(id);
    if (!patient) {
      return res.status(404).json({ error: true, detail: 'Patient not found' });
    }

    // Clean update fields
    for (const key of ['name', 'phone', 'email', 'address']) {
      if (updateFields[key] !== undefined) {
        updateFields[key] = String(updateFields[key]).trim();
      }
    }

    // Duplicate check if name or phone are changing
    if (updateFields.name || updateFields.phone) {
      const checkName = updateFields.name || patient.name;
      const checkPhone = updateFields.phone || patient.phone;

      const dup = await Patient.findOne({
        _id: { $ne: id },
        name: checkName,
        phone: checkPhone,
      });

      if (dup) {
        return res.status(409).json({
          error: true,
          detail: 'A patient with this name and phone already exists',
        });
      }
    }

    // Apply updates
    Object.assign(patient, updateFields);
    await patient.save();
    res.json(patient);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: true, detail: 'Internal server error' });
  }
});

// Delete patient & cascade appointments - Admin only
router.delete('/:id', authenticateJWT, requireRole('admin'), async (req, res) => {
  const { id } = req.params;

  try {
    const patient = await Patient.findById(id);
    if (!patient) {
      return res.status(404).json({ error: true, detail: 'Patient not found' });
    }

    // Cascade delete appointments
    await Appointment.deleteMany({ patient_id: id });
    await Patient.findByIdAndDelete(id);

    res.status(204).send(); // No Content
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: true, detail: 'Internal server error' });
  }
});

module.exports = router;
